### PR TITLE
feat(frontend): streaming support for AI Agent thread responses

### PR DESCRIPTION
**Pull request by notgiorgi**
Closes: #15360

### Description:

Implemented real-time streaming for AI agent responses in the chat interface. This enhancement replaces the static "Thinking..." message with a live-updating response that shows the AI's output as it's being generated.

Key improvements:

- Added a streaming store provider to manage streaming state across components
- Created a dedicated ScrollToBottom component that intelligently handles auto-scrolling during streaming
- Refactored API calls to support streaming responses
- Implemented optimistic UI updates with proper loading states
- Improved the user experience by showing real-time typing feedback instead of a static loading indicator

### CODE REVIEW

**Comment by notgiorgi** [comment]
* **#15515** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15515?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15482** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15482?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15481** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15481?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15480** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15480?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15480?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15479** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15479?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15478** <a href="https://app.graphite.dev/github/pr/lightdash/lightdash/15478?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by tatianainama — COMMENTED** [review]
The conversation list (admin view) is not working, seems like we are missing the new provider in that view

**Comment by notgiorgi** [comment]
@tatianainama good catch! will test that part

**Review by tatianainama — APPROVED** [review]
I ran into the following issue, looks like an edge case so we could work on this later.
1. Start a new thread
2. As soon as streaming starts, navigate to Admin conversation list
3. I can see the streaming in the conversation list
4. Go to thread, thread is empty
5. I cannot see the reply unless I refresh the page

[Screen Recording 2025-06-24 at 13.32.45.mov <span class="graphite__hidden">(uploaded via Graphite)</span> ](https://app.graphite.dev/media/video/HorFj8gm8QUuuIDE2vgk/ebbf3039-f134-4e46-b99a-c8fb51458bfc.mov)

**Comment by notgiorgi** [comment]
@tatianainama good one, this is kind of a known limitation: https://app.graphite.dev/github/pr/lightdash/lightdash/15479/feat-AI-Agent-stream-state-management#comment-PRRC_kwDOFNL_E86A8FRL

but I think I can fix this particular one if I lift the store provider up so that it's shared between those. not sure if this two pages share a common parent though

**Comment by tatianainama** [comment]
I think we could possibly put it in the `/:projectUuid/ai-agents` root?https://github.com/lightdash/lightdash/blob/main/packages/frontend/src/ee/CommercialRoutes.tsx#L81

**Comment by IrakliJani** [comment]
this can be a separate PR I believe

**Comment by IrakliJani** [comment]
unless just moving the provider solves it

File: packages/frontend/src/ee/features/aiCopilot/components/ChatElements/ScrollToBottom.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { useAiAgentThreadStreamQuery } from '../../streaming/useAiAgentThreadStreamQuery';
+
+const SCROLL_TO_BOTTOM_THRESHOLD = 200;
+
+type ScrollToBottomOptions = {
+    behavior?: 'auto' | 'smooth';
+    checkCurrentScrollPosition?: boolean;
+};
+
+function useAutoScroll(scrollAreaRef: React.RefObject<HTMLDivElement | null>) {
```
Comment (IrakliJani) [inline packages/frontend/src/ee/features/aiCopilot/components/ChatElements/ScrollToBottom.tsx:13]: instead of reviewing this FN im gonna vibe check how it feels 👌

File: packages/frontend/src/ee/features/aiCopilot/utils/thinkingMessageStub.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const OPTIMISTIC_MESSAGE_STUB = '___THINKING_STUB___';
+
+export const getOptimisticMessageStub = (): string => {
+    return OPTIMISTIC_MESSAGE_STUB;
+};
```
Comment (IrakliJani) [inline packages/frontend/src/ee/features/aiCopilot/utils/thinkingMessageStub.ts:5]: I don't follow why this whole stub thing is needed

**Review by IrakliJani — APPROVED** [review]
good stuff, minor question

File: packages/frontend/src/ee/features/aiCopilot/utils/thinkingMessageStub.ts (hunk 2 above)
Comment (notgiorgi) [reply packages/frontend/src/ee/features/aiCopilot/utils/thinkingMessageStub.ts:5]: tl;dr; because of optimistic states

Long answer: 
When a message is sent, we put it into react-query state optimistically so it has no actual ID so this is a hacky way to find which message is being streamed (there's always one per thread maximum)

We can do a small refactor around this in a separate PR

**Comment by notgiorgi** [comment]
> I think we could possibly put it in the `/:projectUuid/ai-agents` root?https://github.com/lightdash/lightdash/blob/main/packages/frontend/src/ee/CommercialRoutes.tsx#L81
> 

that did the trick @tatianainama ✨

**Comment by notgiorgi** [comment]
### Merge activity

* **Jun 24, 1:55 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/lightdash/lightdash/15480).
* **Jun 24, 2:01 PM UTC**: [Graphite](https://app.graphite.dev/github/pr/lightdash/lightdash/15480) rebased this pull request as part of a merge.
* **Jun 24, 2:02 PM UTC**: @notgiorgi merged this pull request with [Graphite](https://app.graphite.dev/github/pr/lightdash/lightdash/15480).
